### PR TITLE
[Feature] Add OptimizerHook

### PR DESCRIPTION
**Pull request by aakhundov**
## Description

The new type of hook, `OptimizerHook`, is added as suggested in #704.

## Motivation and Context

See the suggested change in #704.

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] New feature (non-breaking change which adds core functionality)
- [x] Documentation (**initial** update in the documentation)

Backward compatibility with the existing `optimizer` parameter of `Trainer.__init__` is respected.
Minor breaking change: previously `grad_norm` key in the `losses_td` is now `grad_norm_0`.

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

The change made in the documentation is very initial: `optimizer` hook is just briefly mentioned in the list of possible hooks and the `OptimizerHook` class is added to the "Trainer and hooks" table. Probably, wider documentation coverage of the new hook type is needed.

### CODE REVIEW

**Comment by aakhundov** [comment]
@vmoens As currently implemented, the new hook's signature is different from other hooks: besides accepting (and returning) a `TensorDict` instance, it also accepts the clip norm parameters and the index `i` in the `trainer._optimizer_ops` list (for distinguishing each hook's `grad_norm_{i}` key added to the returned `TensorDict`).

If we want the optimizer hook to be more similar to other hooks, we could:

1. Add `clip_norm` and `clip_norm_grad` constructor parameters to the `OptimizerHook.__init__`. Then, when creating the "default" optimizer hook from its `optimizer` argument, `Trainer.__init__` will just pass its own `clip_norm` and `clip_norm_grad` arguments as is to the `OptimizerHook.__init__`. Bonus: users will be able to set different gradient clipping configuration for each hook separately.

2. Return `grad_norm` value from the `OptimizerHook.__call__` instead of the updated `TensorDict` parameter. Then in `Trainer._optimizer_hook` we could set the `grad_norm` in the `losses_td` with the respective index, without the need to pass the index to the hook call.

3. Move detaching the `losses_td` outside of `Trainer._optimizer_hook`: to be done by the calling `Trainer.optim_steps` right after the `Trainer._optimizer_hook` returns.

Importantly, even after the above modifications, `Trainer._optimizer_hook` will be different from other `Trainer._*_hook` functions, as it will need to set `grad_norm` in its `TensorDict` argument besides the usual for-loop over the hook instances.

Please let me know how you'd like to proceed. Thanks!

**Review by vmoens — APPROVED** [review]
Great stuff!
Thanks a lot for this!

**Comment by vmoens** [comment]
Regarding this

> Importantly, even after the above modifications, Trainer._optimizer_hook will be different from other Trainer._*_hook functions, as it will need to set grad_norm in its TensorDict argument besides the usual for-loop over the hook instances.

It's ok, it's backward compatible and it suits most needs.
Thanks again for this!
